modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderFactory.java

```diff
@@ -0,0 +1,246 @@
+    int cacheExpiration = parseCacheExpirationProperty(properties);
+
+    validateOrganization(organization);
+    validateUrl(urlStr);
+    validateSystemUserKey(systemUserKey);
+    validateSystemUserId(systemUserId);
+    validateApplicationId(applicationId);
+    validateApplicationKey(applicationKey);
+
+    ServiceRegistration existingRegistration = this.providerRegistrations.remove(pid);
+    if (existingRegistration != null) {
+      existingRegistration.unregister();
+    }
+
+    Organization org;
+    try {
+      org = this.orgDirectory.getOrganization(organization);
+    } catch (NotFoundException nfe) {
+      logger.warn("Organization {} not found!", organization);
+      throw new ConfigurationException(ORGANIZATION_KEY, "not found");
+    }
+
+    logger.debug("creating new MoodleUserProviderInstance for pid=" + pid);
+
+
+    BrightspaceUserProviderInstance provider = new BrightspaceUserProviderInstance(pid,
+            new BrightspaceClientImpl(urlStr, applicationId, applicationKey, systemUserId, systemUserKey), org, cacheSize,
+            cacheExpiration);
+    this.providerRegistrations
+            .put(pid, this.bundleContext.registerService(UserProvider.class.getName(), provider, null));
+    this.providerRegistrations
+            .put(pid, this.bundleContext.registerService(RoleProvider.class.getName(), provider, null));
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.osgi.service.cm.ManagedServiceFactory#deleted(java.lang.String)
+   */
+  @Override
+  public void deleted(String pid) {
+    logger.debug("delete BrightspaceUserProviderInstance for pid=" + pid);
+    ServiceRegistration registration = providerRegistrations.remove(pid);
+    if (registration != null) {
+      registration.unregister();
+
+      try {
+        ManagementFactory.getPlatformMBeanServer().unregisterMBean(getObjectName(pid));
+      } catch (Exception var4) {
+        logger.warn("Unable to unregister mbean for pid='{}': {}", pid, var4.getMessage());
+      }
+    }
+  }
+
+  private int parseCacheExpirationProperty(Dictionary properties) {
+    try {
+      if (properties.get(CACHE_EXPIRATION) != null) {
+        return Integer.parseInt(properties.get(CACHE_EXPIRATION).toString());
+      }
+    } catch (NumberFormatException nfe) {
+      logger.warn("{} could not be loaded, default value is used: {}", CACHE_EXPIRATION, DEFAULT_CACHE_EXPIRATION_VALUE);
+    }
+    return DEFAULT_CACHE_EXPIRATION_VALUE;
+  }
+
+  private int parseCacheSizeProperty(Dictionary properties) {
+    try {
+      if (properties.get(CACHE_SIZE) != null) {
+        return  Integer.parseInt((String) properties.get(CACHE_SIZE));
+      }
+    } catch (NumberFormatException nfe) {
+      logger.warn("{} could not be loaded, default value is used: {}", CACHE_SIZE, DEFAULT_CACHE_SIZE_VALUE);
+    }
+    return DEFAULT_CACHE_SIZE_VALUE;
+  }
+
+  private void validateSystemUserKey(String userKey) throws ConfigurationException {
```

Comment:
The following four methods are identical except for the key which you could just pass along to the method as arhument.